### PR TITLE
Add rstest as test framework and reconstruct tests with fixtures

### PR DESCRIPTION
**Pull request by bronvic**
#33 

New test framwork rstest used
With it I wrote a lot of fixtures... maybe too much xD

### CODE REVIEW

File: backend/src/fixtures/grid.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+pub fn col_n() -> u32 {
+    4
+}
+
+#[fixture]
+pub fn grid() -> Grid {
+    Grid::new(row_n(), col_n())
+}
+
+#[fixture()]
```
[the remark this answers is not in the document]
Comment (alex-eg) [reply backend/src/fixtures/grid.rs:19]: However, it's not very useful to have separate fixtures for each field value, as later, when we actually will have some our code to test, we would want to test corner cases and complex behaviours, and fixtures for individual field values will not help us with this.
Consider the function `attack_unit(u1: Unit, u2: Unit) -> Unit`, and how we will want to test it. Rstest has a nice mechanism to pass arbitrary parameters to the test
```Rust
pub fn attack_unit(u1: Unit, u2: Unit) -> Unit {
    Unit {
        hp: if u1.attack > u2.hp {
            0
        } else {
            u2.hp - u1.attack
        },
        ..u2
    }
}

mod test {
    use super::*;

    fn test_unit() -> Unit {
        Unit {
            player: 0,
            hp: 22,
            attack: 10,
            speed: 3,
        }
    }

    #[rstest(
        hp,
        attack,
        result_hp,
        case::reduce_hp(100, 99, 1),
        case::reduce_hp(2, 1, 1),
        case::reduce_hp(10, 0, 10),
        case::kill(10, 10, 0),
        case::kill(10, 100, 0),
        case::kill(0, 10, 0),
    )]
    fn test_attack(attack: u32, hp: u32, result_hp: u32) {
        let u1 = Unit {
            attack,
            ..test_unit()
        };
        let u2 = Unit { hp, ..test_unit() };
        let u3 = attack_unit(u1, u2);
        assert_eq!(u3.hp, result_hp);
    }
}
```

[the remark this answers is not in the document]
Comment (alex-eg) [reply backend/src/fixtures/grid.rs:19]: Also, fixtures, as name suggests, are mainly used for setting up testing environment, which usually involves some complex operations, like opening database connection, copying files in some directory, or creating a number of mock objects. In either way, it mainly concerns things which doesn't change much between tests, and thus it isn't well suited for testing objects. We usually want our test data to be easily parametrized, as good tests covers as many input data combinations, as possible.

**Comment by bronvic** [comment]
rotten
